### DISCUSSION TITLE
Feature Request: Parallel Computation/Processing?

### DISCUSSION THREAD

**Discussion by EkaterinaOssetchkina** [pyvista/pyvista#1537] [Ideas]
Hello! We are working with several GB of models (individual vtk files), which are over many timesteps. What we are trying to do is extract information from several arrays, at each time step, and then append the information from each time step into a new array.

This takes a while to process, and we have tried to integrate PyVista with parallel threading/parallel processing libraries, but we are not getting much time savings/results:
- **threading library** - would not execute
- **multiprocessing library** - was not successful in mapping processes to our thresholding function over the timesteps, or would not return any results.
- **concurrent.futures**
          -**Process Pools** - executed, but only savings of 10 seconds on a 12 GB model, we also noticed that although several processes are started, only about a quarter of them seem to be doing the processing work (example, 2/8 processes would work)
          -**Thread Pools** - executed, but only savings of ~10 seconds on a 12 GB model as well, same as Process Pools
- **numba with jit** - got error messages that it did not recognize PyVista data array types 

This is a follow up of #1401 from a while ago. 
We were wondering what is the recommended approach with PyVista for this type of problem, or if anything was added to PyVista for Parallel Processing I/O. Thank you!

**Comment by RichardScottOZ** [comment]
12 GB is one file or many?

**Comment by whophil** [comment]
Hi @EkaterinaOssetchkina , more detail would be helpful.

How do you select which data you would like to extract into a new array? Do you need to run filters? How large is one individual data file?

Regarding process parallelization with `concurrent.futures` or `multiprocessing` - your performance can vary widely depending on how the parallelism has been implemented.

Passing PyVista objects through `ProcessPoolExecutor`, which requires serialization/deserialization through a VTK file format, has some overhead (see https://github.com/pyvista/pyvista/pull/1778). You may already be doing this, but it is likely faster to read the files within the individual executors and return the required subsets of data - either as PyVista objetcts or as NumPy arrays. (i.e., map the filenames, rather than the PyVista data, onto the executors and return the data subsets).

If you're truly I/O bound, I don't think there is much you can do. It's probably worth profiling your process in serial to see what the bottlenecks are.
